src/registrar/admin.py

```diff
@@ -1478,6 +1484,120 @@ def changelist_view(self, request, extra_context=None):
+
+            requested_user = User.objects.filter(email=requested_email).first()
+            permission_exists = UserPortfolioPermission.objects.filter(
+                user=requested_user, portfolio=portfolio
+            ).exists()
+            try:
+                if not requested_user or not permission_exists:
+                    # if requested user does not exist or permission does not exist, send email
+                    send_portfolio_invitation_email(email=requested_email, requestor=requestor, portfolio=portfolio)
+                    messages.success(request, f"{requested_email} has been invited.")
+                elif permission_exists:
+                    messages.warning(request, "User is already a member of this portfolio.")
+            except Exception as e:
+                # when exception is raised, handle and do not save the model
+                self._handle_exceptions(e, request, obj)
+                return
+        # Call the parent save method to save the object
+        super().save_model(request, obj, form, change)
+
+    def _handle_exceptions(self, exception, request, obj):
+        """Handle exceptions raised during the process.
+
+        Log warnings / errors, and message errors to the user.
+        """
+        if isinstance(exception, EmailSendingError):
+            logger.warning(
+                "Could not sent email invitation to %s for portfolio %s (EmailSendingError)",
+                obj.email,
+                obj.portfolio,
+                exc_info=True,
+            )
+            messages.error(request, "Could not send email invitation. Portfolio invitation not saved.")
+        elif isinstance(exception, MissingEmailError):
+            messages.error(request, str(exception))
+            logger.error(
+                f"Can't send email to '{obj.email}' for portfolio '{obj.portfolio}'. "
+                f"No email exists for the requestor.",
+                exc_info=True,
+            )
+
+        else:
+            logger.warning("Could not send email invitation (Other Exception)", exc_info=True)
+            messages.error(request, "Could not send email invitation. Portfolio invitation not saved.")
+
+    def response_add(self, request, obj, post_url_continue=None):
+        """
+        Override response_add to handle rendering when exceptions are raised during add model.
+
+        Normal flow on successful save_model on add is to redirect to changelist_view.
+        If there are errors, flow is modified to instead render change form.
+        """
+        # Check if there are any error or warning messages in the `messages` framework
+        storage = get_messages(request)
+        has_errors = any(message.level_tag in ["error", "warning"] for message in storage)
+
+        if has_errors:
+            # Re-render the change form if there are errors or warnings
+            # Prepare context for rendering the change form
+            opts = self.model._meta
```

Comment:
**non blocking question** why "opts"?

Reply:
If confusing, can change. Was modeling code after super class, django.contrib.admin.options.ModelAdmin, in which many of their methods get the options from _meta.